### ISSUE TITLE
Getting Error: "There was an error sending message to your bot. HTTP status code Unauthorised"

### ISSUE THREAD

**Issue by ashwani-ksprof** [Microsoft/BotBuilder#4269]
## Bot Info
* SDK Platform: .NET
* SDK Version: Bot Builder SDK 3.9.0.0
* Active Channels: Directline and WebChat
* Deployment Environment: In-house Server

AppID = aa0cd191-5aea-4c4a-8fad-73f046e65fe6
BotHandle = BrianStaging

## Issue Description
I am getting below error all of a sudden on a running bot.

![image](https://user-images.githubusercontent.com/28593146/37088764-72e01454-2224-11e8-8a6d-593946450450.png)

I have deployed the Bot solution on to our in house server with valid SSL certificate.
Then endpoint of the API is https://brian.birlasoft.com/uat/api/messages 

The same app was working till 28 Feb 2018 and now it is giving the above error when using Dirctline and WebChat channels.

But, when we try to connect through Emulator then it works seems like emulator directly connect to Bot API bypassing the channel connector.

Please help me troubleshooting the issue as we are stuck here.

Here is the Directline secret key we are using to connect to the bot.
[mod: RemovedSecret]

**Comment by Aashish-Amber** [comment]
Same thing happened with multiple of my bots , try re-generating your app ID and password and use the new one instead of the old , it will solve the problem.

**Comment by kamidipreetham** [comment]
@Aashish-Amber Just the app password is enough. On the bot side this error shows up: **ChatConnector: receive - invalid signing key or OpenId metadata document**.

**Comment by ashwani-ksprof** [comment]
Thanks for the suggestions.
However, I have already created a new app with new set of AppId and Password, even then it is not working.
As I mentioned in the issue description that I have been able to connect to my Bot using emulator which require to pass AppId and Password. So AppId and Password if working but seems like issue with Directline and webchat channel.

Any thoughts around this.

**Comment by dvorakjan** [comment]
The same problem with `ChatConnector: receive - invalid signing key or OpenId metadata document.` on our multiple bot instances. It fails randomly, not on every request. Error appears for requestst sent at least through directline, webchat and facebook channels. Restart of bot solves problem for a while.

**Comment by Kannaj** [comment]
+1 - facing the same issue since yesterday night. Heres the error we're getting on the bot side

`ERROR: ChatConnector: receive - invalid signing key or OpenId metadata document.`
Our MSA ID - 71fad264-9fd6-4ecd-808d-becdbc3b7eb1

Our SSL certs seem to be fine. 

our botbuilder version is 3.11 .

**Comment by kamidipreetham** [comment]
@ashwani-ksprof The issue occurs for facebook channels too, I don't think it's limited to webchat/directline.

**Comment by james60338** [comment]
Not sure whether I faced the same issue, too, tried for several hours.
- My text message function on Skype bot worked fine, but when I call to the bot it replies "you can't talk to this bot just yet but we are working on it", and then hung up.
- Tried to debug by ngrok, but my controller was not even been called, the error is : 401 Unauthorized, "BotAuthenticator failed to authenticate incoming request!"
- Also tried to build new bot and new Skype bot channel but didn't fix the issue.

**Comment by stuartd** [comment]
We have this on the facebook channel, and intermittently on our homebrew directline implementation.

Looking at our logs, message are received and processed by the framework, however the response is not delivered.

![image](https://user-images.githubusercontent.com/275326/37100587-453c819e-221b-11e8-8894-52b6dbd2c5ad.png)

```
{
	"incoming": {
		"type": "message",
		"id": "xxx",
		"timestamp": "2018-03-07T15:19:22.017Z",
		"serviceUrl": "https://facebook.botframework.com",
		"channelId": "facebook",
		"from": {
			"id": "1136148923120997",
			"name": "stuartd"
		},
		"conversation": {
			"isGroup": false,
			"id": "xxx"
		},
		"recipient": {
			"id": "xxx",
			"name": "xxx"
		},
		"membersAdded": [],
		"membersRemoved": [],
		"text": "test for issue 4269",
		"attachments": [],
		"entities": [],
		"channelData": {
			"sender": {
				"id": "xxx"
			},
			"recipient": {
				"id": "xxx"
			},
			"timestamp": 1520435962017,
			"message": {
				"mid": "mxxx",
				"seq": 117197,
				"text": "test for issue 4269",
				"is_echo": false
			}
		}
	}
}
```

```
{
	"outgoing": {
		"type": "message",
		"serviceUrl": "https://facebook.botframework.com",
		"channelId": "facebook",
		"from": {
			"id": "xxx",
			"name": "xxx"
		},
		"conversation": {
			"isGroup": false,
			"id": "xxx"
		},
		"recipient": {
			"id": "xxx"
		},
		"membersAdded": [],
		"membersRemoved": [],
		"reactionsAdded": [],
		"reactionsRemoved": [],
		"text": "Sorry, I don't understand what you mean.{linebreak}",
		"attachments": [],
		"entities": [],
		"replyToId": "xxx"
	}
}
```

One of our team in is in SF and supposed to be doing demos all week..

**Comment by JasonSowers** [comment]
Please try restarting your bot service and see if that fixes the issue.

**Comment by dvorakjan** [comment]
Restart of bot service fixed issue (now its  more than 7 hour since restart) for all our bots. But we are afraid that it can come back again since we don’t know the reason. Should we use some other/faster channel than Github in case of future problems?

**Comment by EricDahlvang** [comment]
Hi @DarkInsight85 

The official solution is to restart the bot:

![image](https://user-images.githubusercontent.com/11055362/37132639-879454cc-2243-11e8-8b23-74b49c15aa56.png)

**Comment by EricDahlvang** [comment]
@ashwani-ksprof please try upgrading the bot builder sdk.

**Comment by ashwani-ksprof** [comment]
@EricDahlvang I have upgraded my bot with SDK version 3.13.1. But I am still getting the same issue.
Please guide me how to resolve this as we had to make this bot live for some pilot user for testing and has already been delayed by one week.

**Comment by EricDahlvang** [comment]
@ashwani-ksprof  

Your issue appears to be different from the others in this thread.  Have you tried setting up ngrok, and changing the messaging endpoint, and debugging locally?  https://blog.botframework.com/2017/10/19/debug-channel-locally-using-ngrok/

If you can share your project with me, please email it to the address on my GitHub profile: https://github.com/EricDahlvang

**Comment by ashwani-ksprof** [comment]
@EricDahlvang 
Yes, I did tried with ngrok and debugged the code locally for directline and web chat channel. It is working perfectly.

I am not sure on sharing the code as it would be a compliant issue as suggested by seniors in my team.
Let me check again if I could share the code, meanwhile, if you can suggest or have any other idea to fix the issue, please let me know.

**Comment by ashwani-ksprof** [comment]
What else, then, should be the messaging endpoint? This is what we configure in the bot app on Azure.

![image](https://user-images.githubusercontent.com/28593146/37325515-3bfa265c-26b4-11e8-8836-b13084c97973.png)

We have been running this bot for the last 9 months and it was working just fine.
We have developed our own client which is a mobile app (android) using DirectLine channel. We are using DirectLine secret key for communicating with the bot. 
Yes, we are also using WebChat channel to just plugin the bot in out intranet web site, but not using WebChat iframe, instead, we are using botchat.js to initiate the conversation with the bot like below:

  <script>            
        var email = 'ashwani1.kumar@BIRLASOFT.COM';
        var userName = 'ashwani';        
        var model = {
            "userId": email,
            "userName": userName,
            "botIconUrl": "~/images/user.jpg",            
            "secret": "*********-PQc.********.Ber3I6Bu_*****************************",
            "iconUrl": "~/images/user.jpg",
            "directLineUrl": "https://webchat.botframework.com/v3/directline",
            "webSocketEnabled": "false"
        };
 </script>
 <script>
        BotChat.App({
            directLine: {
                secret: model.secret,
                token: model.token,
                domain: model.directLineUrl,
                webSocket: false
            },
            user: { id: model.userId, name: model.userName },
            bot: { id: model.botId, name: model.botName },
            resize: 'window'
        }, document.getElementById("BotChatElement"));
 </script>

Any other information I can help you with, please let me know.

**Comment by EricDahlvang** [comment]
@ashwani-ksprof   

It is not clear to me why your bot is experiencing this issue.  When you say you've setup ngrok, and debugged locally, did you do this by changing the messaging endpoint in Azure to point to the ngrok endpoint?  Then use webchat, or your custom directline client, to connect to the bot?  (i'm not referring to using the emulator)

You could also try regenerating your secrets.  (Though I've never seen this solve anyone's issue before, I've also never seen this unexplained authorization exception.)

Lastly, why are you using 'https://webchat.botframework.com/v3/directline' for the directLineUrl above?  I think it should be https://directline.botframework.com/v3/directline reference: https://github.com/Microsoft/BotFramework-DirectLineJS/blob/af5afc7e9dd4cb7ef5f50ae3ee9881f939556467/src/directLine.ts#L287

**Comment by EricDahlvang** [comment]
@ashwani-ksprof Yes, we can setup a time to do a skype screenshare.  Please email me at the address on my profile: https://github.com/EricDahlvang

If you could create a sample project that demonstrates this issue, it would be helpful.  I don't need your secrets or passwords, but just a small project that I can use to recreate the issue.

**Comment by ashwani-ksprof** [comment]
@EricDahlvang 
I have sent you a mail. Please acknowledge and let me know your availability.

**Comment by padmavatierp** [comment]
i am also dealing with the same issue, is there any solution pls let me know
thanks in advance

**Comment by EricDahlvang** [comment]
@padmavatierp  Please open a new issue.

**Comment by padmavatierp** [comment]
HI eric some thing wrong with my bot
can u pls contact me on skype?

**Comment by suresh1271** [comment]
Hi @EricDahlvang  I am having the same issue as @ashwani-ksprof  mentioned. Can you please help in this. 

@ashwani-ksprof  hope you might got the solution for this problem. If yes, please share the details to resolve.

Thanks!!

**Comment by ashwani-ksprof** [comment]
> Hi @EricDahlvang I am having the same issue as @ashwani-ksprof mentioned. Can you please help in this.
> 
> @ashwani-ksprof hope you might got the solution for this problem. If yes, please share the details to resolve.
> 
> Thanks!!

Hi @suresh1271 
In my case, the issue was occurring due to mismatch of date time between the server where I published my bot and the channel connector on azure for my bot.

Whenever any message was sent to my bot from channel connector through web chat or directline, it says unauthorized just because the token generated for the time period was not yet valid as my server (_which was actually set to the correct time as per the timezone_) was approx. 6 min. behind from the time token was generated.
So, I had to explicitly set my server clock 6 min ahead of the current time and then it worked.

It was really a weird issue and you might also be facing the same.

Hope it helps.

Thanks,
Ashwani Kumar
